File: rails/app/data/deck_shoe.rb
# frozen_string_literal: true

class DeckShoe < Deck
  attr_accessor :card_on_top

  @card_on_top

  # @param [Integer] num_decks number of decks in the shoe
  def initialize(num_decks)
    super()
    (num_decks - 1).times do
      add_default_cards
    end
    shuffle
  end

  def shuffle
    @card_on_top = 0
    super
  end

  def next_card
    shuffle if @card_on_top == @cards.length
    card = @cards[@card_on_top]
    @card_on_top = @card_on_top + 1
    card
  end
end

File: django/blackjack/store/deck_shoe.py
from blackjack.state.card import Card
from blackjack.store.deck import Deck

class DeckShoe(Deck):
    def __init__(self, numDecks: int) -> None:
        super().__init__()
        for _ in range(0, numDecks - 1):
            self.addDefaultCards()
        self.shuffle()

    def shuffle(self):
        ret = super().shuffle()
        self.cardOnTop = 0
        return ret

    def getNextCard(self) -> Card:
        if self.cardOnTop >= len(self.cards):
            self.shuffle()
        card = self.cards[self.cardOnTop]
        self.cardOnTop = self.cardOnTop + 1
        return card
